esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6448,36 +6481,205 @@ bool CWsDfuEx::onDFUFilePublish(IEspContext &context, IEspDFUFilePublishRequest
+    }
+    catch (IException *e)
+    {
+        exception.setown(e);
+    }
+
+    if (exception)
+    {
+        if (fileDesc)
+        {
+            Owned<IMultiException> exceptions = MakeMultiException("CWsDfuEx::onDFUFilePublish");
+            queryDistributedFileDirectory().removePhysicalPartFiles(normalizeTempFileName, fileDesc, exceptions);
+            if (exceptions->ordinality())
+            {
+                StringBuffer errMsg("Error whilst clearing up temporary file: ");
+                EXCLOG(exceptions, errMsg.append(normalizeTempFileName).str());
+            }
+        }
+        FORWARDEXCEPTION(context, exception.getClear(), ECLWATCH_INTERNAL_ERROR);
+    }
+    return true;
+}
+
+bool CWsDfuEx::onDFUPublishFile(IEspContext &context, IEspDFUPublishFileRequest &req, IEspDFUFilePublishResponse &resp)
+{
+    Owned<IException> exception;
+    Owned<IFileDescriptor> fileDesc;
+    StringBuffer normalizeTempFileName;
+    try
+    {
+        context.ensureFeatureAccess(FEATURE_URL, SecAccess_Write, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUPublishFile: Permission denied.");
+
+        const char *fileId = req.getFileId();
+        if (isEmptyString(fileId))
+             throw makeStringException(ECLWATCH_INVALID_INPUT, "No FileId defined.");
+
+        const char *recordDefinition = req.getECLRecordDefinition();
+
+        CDfsLogicalFileName lfn;
+        const char *tempFileName = nullptr;
+        StringBuffer groupName;
+
+        StringArray fileIdItems;
+        fileIdItems.appendList(fileId, DFUFileIdSeparator);
+
+        // Either group|cluster|tempFileName OR fileDesc provided and fileId=tempFileName
+        if (fileIdItems.ordinality() < 3)
+        {
+            // NB: has to be >=1 as already checked not empty
+            if (1 == fileIdItems.ordinality())
+            {
+                const MemoryBuffer &fDescBlob = req.getFileDescriptorBlob();
+                if (fDescBlob.length())
+                {
+                    MemoryBuffer mb;
+                    mb.setBuffer(fDescBlob.length(), (void *)fDescBlob.toByteArray());
+                    fileDesc.setown(deserializeFileDescriptor(mb));
+                }
+            }
+            if (!fileDesc)
+                throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid FileId '%s'. or missing file descriptor", fileId);
+            tempFileName = fileIdItems.item(0);
+            lfn.set(tempFileName);
+            normalizeTempFileName.set(lfn.get());
+            fileDesc->getClusterGroupName(0, groupName);
+        }
+        else
+        {
+            groupName.set(fileIdItems.item(0));
+            const char *clusterName = fileIdItems.item(1);
+            tempFileName = fileIdItems.item(2);
+            if (isEmptyString(groupName))
+                 throw makeStringException(ECLWATCH_INVALID_INPUT, "Invalid FileId: empty groupName.");
```

Comment:
let's provide method name in all err messages for context

Reply:
agreed, will change.